### DISCUSSION TITLE
Question: Missing package.json file when releasing a package with package.json in a subfolder

### DISCUSSION THREAD

**Discussion by SolidAlloy** [openupm/openupm#1875] [Q&A] [answered]
There is a package already published to OpenUPM: https://openupm.com/packages/com.madsbangh.easybuttons/
I am making a PR to the repository trying to apply semantic versioning so that the most recent version of the package is always available on OpenUPM.

I followed your tutorial on how to apply SemVer for repositories with a UPM branch https://medium.com/openupm/how-to-maintain-upm-package-part-2-f352fbf5f87c#6bf4

Before publishing the PR, I would like to test it on the fork but npm publishing is failing with the following error:
"Missing `package.json` file."

Here's an issue created by Semver in my fork: https://github.com/SolidAlloy/EasyButtons/issues/1

I checked paths multiple times and made sure they are correct. The package.json check also fails when I do a dry run on a local machine.

I can see a similar error in the [Build Pipelines](https://openupm.com/packages/com.madsbangh.easybuttons/?subPage=pipelines) of the package on the OpenUPM website.

Do you have any idea what can be wrong with this setup? Or is it somehow happening because the repo is a fork?
Any help will be appreciated.

**Comment by favoyang** [comment, marked as answer]
Hi @SolidAlloy,

Moved the thread from issue to discussion, because this is a question.

The message `error path package.json` means the `npm publish` can not locate the package.json file in the current folder (of a given Git tag).

https://openupm.com/packages/com.madsbangh.easybuttons/?subPage=pipelines
- It means none of the existing tags has a package.json file at any path. Those tags are very old and before the author adds the upm support.
- The author also has some custom build script that creates a upm branch, but the package.json version is still `1.0.0`.
- The author should
  - Bump the package.json version, perhaps to v1.1.1.
  - Create Git tag from the master, then create the second Git tag from the upm branch (to serve OpenUPM). i.e. `v1.1.1` from the main branch, then `upm/v1.1.1` for the upm branch. If the author uses other prefixes than `upm/`, needs to edit [the YAML package file's `gitTagPrefix` field](https://github.com/openupm/openupm/blob/master/data/packages/com.madsbangh.easybuttons.yml) to inform OpenUPM build-pipelines.

**Reply by favoyang** [reply]
@SolidAlloy,

In your fork, you try to apply the tutorial by
1. use semantic-release-action@v2 to automate the release.
2. create a separate upm branch.
3. create a versioned Git tag from the upm branch (if package.json version changed).

Step 1 you use makes an assumption that package.json is located at the root path unless you set the `pkgRoot` configuration https://github.com/favoyang/semantic-release-upm-example/blob/master/.releaserc.json. That's why it failed in https://github.com/SolidAlloy/EasyButtons/issues/1.

Just my 2 cents, applying semantic-release is a good practice, but because it's a fork, you need to discuss it with the original authors whether they like it or not. Otherwise, if you publish a fork on OpenUPM and the version keeping goes up, then the original author may not like it when starts to manage the version. Anyway managing a fork version is a bit tricky. https://medium.com/openupm/how-to-maintain-upm-package-part-3-2d08294269ad.

One workaround is that you change the package naming to your namespace, then you can apply any version management strategy to it.

**Comment by SolidAlloy** [comment]
@favoyang I didn't intend to publish the fork on OpenUPM. Just wanted to set SemVer for the original repo and submit it as a PR. I somehow missed that .releaserc.json needs to be at the repo root. I had it in the same folder as package.json, and once I moved it to the root, the release automation started working. Thanks for the detailed answer!

**Reply by favoyang** [reply]
Got it. Then you're preparing the PR of the automatic release.
